File: frontend/src/api/modules/control-plane.ts
import { request, setStoredSessionToken } from '../core'

export type SavedViewScope = 'sources' | 'alerts' | 'artifacts' | 'history'

export interface Workspace {
  id: string
  name: string
  slug: string
  description?: string | null
  is_default: boolean
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface Role {
  id: string
  name: string
  description?: string | null
  permissions: string[]
  is_system: boolean
  created_at: string
  updated_at: string
}

export interface Permission {
  id: string
  key: string
  category: string
  description?: string | null
  is_system: boolean
  created_at: string
  updated_at: string
}

export interface User {
  id: string
  email: string
  display_name: string
  is_active: boolean
  is_system: boolean
  preferences: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface SessionContext {
  token?: string | null
  expires_at?: string | null
  user: User
  workspace: Workspace
  role: Role
}

export interface SavedView {
  id: string
  scope: SavedViewScope
  name: string
  description?: string | null
  filters: Record<string, unknown>
  is_default: boolean
  owner_id: string
  owner_name?: string | null
  workspace_id: string
  created_at: string
  updated_at: string
}

export interface SavedViewListResponse {
  data: SavedView[]
  total: number
  offset: number
  limit: number
}

export interface OverviewSlice {
  total: number
  active?: number | null
  healthy?: number | null
  unhealthy?: number | null
  failed?: number | null
  fresh_24h?: number | null
  stale?: number | null
  unowned?: number | null
}

export interface OverviewSavedView {
  id: string
  name: string
  scope: SavedViewScope
  description?: string | null
  is_default: boolean
  owner_name?: string | null
}

export interface OverviewQueueBacklog {
  queue_id: string
  queue_name: string
  count: number
}

export interface OverviewAssigneeLoad {
  user_id?: string | null
  user_name: string
  count: number
}

export interface OverviewArtifactTypeCount {
  artifact_type: string
  count: number
}

export interface OverviewNamedCount {
  id?: string | null
  name: string
  count: number
}

export interface OverviewOwnershipFreshness {
  ownership_type: string
  ownership_id?: string | null
  ownership_name: string
  fresh_24h: number
  stale: number
}

export interface OverviewWorkspaceSummary {
  id: string
  name?: string | null
  slug?: string | null
}

export interface OverviewResponse {
  workspace: OverviewWorkspaceSummary
  sources: OverviewSlice
  incidents: OverviewSlice
  artifacts: OverviewSlice
  incident_backlog: OverviewQueueBacklog[]
  assignee_workload: OverviewAssigneeLoad[]
  artifact_types: OverviewArtifactTypeCount[]
  sources_by_owner: OverviewNamedCount[]
  sources_by_team: OverviewNamedCount[]
  sources_by_domain: OverviewNamedCount[]
  artifact_freshness_by_ownership: OverviewOwnershipFreshness[]
  saved_views: OverviewSavedView[]
}

export async function getSession(): Promise<SessionContext> {
  const session = await request<SessionContext>('/auth/session')
  if (session.token) {
    setStoredSessionToken(session.token)
  }
  return session
}

export async function createSession(password?: string, workspace_id?: string): Promise<SessionContext> {
  const session = await request<SessionContext>('/auth/session', {
    method: 'POST',
    body: JSON.stringify({ password, workspace_id }),
  })
  if (session.token) {
    setStoredSessionToken(session.token)
  }
  return session
}

export async function logoutSession(): Promise<void> {
  await request<{ message: string }>('/auth/session', { method: 'DELETE' })
  setStoredSessionToken(null)
}

export async function getCurrentUser(): Promise<SessionContext> {
  return request<SessionContext>('/me')
}

export async function listWorkspaces(): Promise<Workspace[]> {
  return request<Workspace[]>('/workspaces')
}

export async function listUsers(): Promise<User[]> {
  return request<User[]>('/users')
}

export async function listRoles(): Promise<Role[]> {
  return request<Role[]>('/roles')
}

export async function listPermissions(): Promise<Permission[]> {
  return request<Permission[]>('/permissions')
}

export async function listSavedViews(scope?: SavedViewScope): Promise<SavedViewListResponse> {
  return request<SavedViewListResponse>('/views', {
    params: scope ? { scope } : undefined,
  })
}

export async function createSavedView(payload: {
  scope: SavedViewScope
  name: string
  description?: string
  filters: Record<string, unknown>
  is_default?: boolean
}): Promise<SavedView> {
  return request<SavedView>('/views', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export async function updateSavedView(
  viewId: string,
  payload: {
    name?: string
    description?: string
    filters?: Record<string, unknown>
    is_default?: boolean
  }
): Promise<SavedView> {
  return request<SavedView>(`/views/${viewId}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

export async function deleteSavedView(viewId: string): Promise<void> {
  await request<{ message: string }>(`/views/${viewId}`, { method: 'DELETE' })
}

export async function getOverview(): Promise<OverviewResponse> {
  return request<OverviewResponse>('/overview')
}

File: src/truthound_dashboard/schemas/control_plane.py
"""Schemas for dashboard-native control-plane APIs."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import Field

from .base import BaseSchema, IDMixin, ListResponseWrapper, TimestampMixin


class RoleResponse(BaseSchema, IDMixin, TimestampMixin):
    name: str
    description: str | None = None
    permissions: list[str] = Field(default_factory=list)
    is_system: bool = True


class PermissionResponse(BaseSchema, IDMixin, TimestampMixin):
    key: str
    category: str
    description: str | None = None
    is_system: bool = True


class WorkspaceResponse(BaseSchema, IDMixin, TimestampMixin):
    name: str
    slug: str
    description: str | None = None
    is_default: bool = False
    is_active: bool = True


class UserResponse(BaseSchema, IDMixin, TimestampMixin):
    email: str
    display_name: str
    is_active: bool = True
    is_system: bool = False
    preferences: dict[str, Any] = Field(default_factory=dict)


class MembershipResponse(BaseSchema, IDMixin, TimestampMixin):
    user: UserResponse
    workspace: WorkspaceResponse
    role: RoleResponse
    is_default: bool = False


class SessionResponse(BaseSchema):
    token: str | None = None
    expires_at: datetime | None = None
    user: UserResponse
    workspace: WorkspaceResponse
    role: RoleResponse


class SessionCreateRequest(BaseSchema):
    password: str | None = None
    workspace_id: str | None = None


SavedViewScope = Literal["sources", "alerts", "artifacts", "history"]


class SavedViewBase(BaseSchema):
    scope: SavedViewScope
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=1000)
    filters: dict[str, Any] = Field(default_factory=dict)
    is_default: bool = False


class SavedViewCreate(SavedViewBase):
    pass


class SavedViewUpdate(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=1000)
    filters: dict[str, Any] | None = None
    is_default: bool | None = None


class SavedViewResponse(SavedViewBase, IDMixin, TimestampMixin):
    owner_id: str
    workspace_id: str
    owner_name: str | None = None


class SavedViewListResponse(ListResponseWrapper[SavedViewResponse]):
    pass


class OverviewSlice(BaseSchema):
    total: int = 0
    active: int | None = None
    healthy: int | None = None
    unhealthy: int | None = None
    failed: int | None = None
    fresh_24h: int | None = None
    stale: int | None = None
    unowned: int | None = None


class OverviewQueueBacklog(BaseSchema):
    queue_id: str
    queue_name: str
    count: int = 0


class OverviewAssigneeLoad(BaseSchema):
    user_id: str | None = None
    user_name: str
    count: int = 0


class OverviewArtifactTypeCount(BaseSchema):
    artifact_type: str
    count: int = 0


class OverviewWorkspaceSummary(BaseSchema):
    id: str
    name: str | None = None
    slug: str | None = None


class OverviewSavedView(BaseSchema):
    id: str
    name: str
    scope: SavedViewScope
    description: str | None = None
    is_default: bool = False
    owner_name: str | None = None


class OverviewNamedCount(BaseSchema):
    id: str | None = None
    name: str
    count: int = 0


class OverviewOwnershipFreshness(BaseSchema):
    ownership_type: str
    ownership_id: str | None = None
    ownership_name: str
    fresh_24h: int = 0
    stale: int = 0


class OverviewResponse(BaseSchema):
    workspace: OverviewWorkspaceSummary
    sources: OverviewSlice
    incidents: OverviewSlice
    artifacts: OverviewSlice
    incident_backlog: list[OverviewQueueBacklog] = Field(default_factory=list)
    assignee_workload: list[OverviewAssigneeLoad] = Field(default_factory=list)
    artifact_types: list[OverviewArtifactTypeCount] = Field(default_factory=list)
    sources_by_owner: list[OverviewNamedCount] = Field(default_factory=list)
    sources_by_team: list[OverviewNamedCount] = Field(default_factory=list)
    sources_by_domain: list[OverviewNamedCount] = Field(default_factory=list)
    artifact_freshness_by_ownership: list[OverviewOwnershipFreshness] = Field(default_factory=list)
    saved_views: list[OverviewSavedView] = Field(default_factory=list)
